### DISCUSSION TITLE
Consider implementing exclusion to refinementList

### DISCUSSION THREAD

**Discussion by xPaw** [algolia/instantsearch#4754] [Ideas]
Spinning out of this old issue: #1232

I still think this is a worthwhile feature to have in instant search out of the box.

I have it implemented here, you can check out how it works there: https://steamdb.info/instantsearch/ (e.g. Tags). It's a bit hacky, but gets the job done.

Having to implement exclusion myself means I need to reinvent most of the code for the widget (e.g. I can't just enabled `searchable` on all the facets).

![image](https://user-images.githubusercontent.com/613331/117259686-f70a4d00-ae56-11eb-98a6-01eb07a5d84d.png)

**Comment by Haroenv** [comment]
The UX for a negative / exclude refinement list is very tough, I haven't yet seen a option that looks fully native (although I do definitely like your idea).

In the mean time we have released a community standalone negative refinement list connector and widget: https://github.com/Haroenv/instantsearch-negative-refinement-list

**Comment by xPaw** [comment]
> The UX for a negative / exclude refinement list is very tough

I agree, and this is partially why I think it would be great if Algolia solved this.

**Comment by xPaw** [comment]
Here's another issue I ran into with my exclusions widget that I think would be possible to solve if implemented properly: if you want to do `OR` searches, but make exclusions still work, they would have to be applied as `AND`. For example `-windows AND (macos OR linux)`.

**Reply by Haroenv** [reply]
That's a good point, and the facetExclusion actually works this way. You'll likely have to edit the refinement list connector to take those refinements in account too though, as the isRefined property will need to be set correctly

**Comment by xPaw** [comment]
I tried using `helper.addFacetExclusion`, it seems like it would be possible to use this in a custom refinementList instead of refining by prepending a minus.
However I am seeing that adding exclusions this way does not update url history router.
Is this intentional, or an oversight?

**Reply by Haroenv** [reply]
Have you checked out the example which does basically that? https://github.com/Haroenv/instantsearch-negative-refinement-list/blob/master/src/lib/connectNegativeRefinementList.ts

**Reply by xPaw** [reply]
Yeah, that's where I got that function from. I was calling it from `connectRefinementList`. It seems that `getWidgetUiState` also needs to be implemented somehow?

Seems a bit unnecessary when `refinementList` itself does, just that exclusions are not propagated to the url, even though they do show up in `currentRefinements`.

**Reply by Haroenv** [reply]
Yes, getWidgetUiState is the function that sets the URL. To be able to be included, you can change the getWidgetUiState and getWidgetSearchParameters of refinement list or a separate widget to include the negative refinements as well.

CurrentRefinements is legacy and based directly on search parameters, but that's not our long term approach, as it doesn't allow for widgets that don't express search parameters

**Reply by xPaw** [reply]
That is a quite a bit of work, hah. Is there a reason Algolia isn't willing to support exclusions in widgets properly?

https://store.steampowered.com/search added tag exclusions not long ago, you can take a peek on how they handle them.

And what is the alternative to CurrentRefinements?

**Reply by Haroenv** [reply]
It would still be CurrentRefinements, but based on UiState in the future (but after a major version, no rush in that)

We have not invested in this yet as there have been very few interest from customers in implementing this pattern

**Reply by xPaw** [reply]
Am I missing something, or connectRefinementList only allows me to change the render function? If I wanted to fix up getWidgetUiState and getWidgetSearchParameters, I would have to essentially reimplement the whole widget?

**Reply by Haroenv** [reply]
Yes, if you are adding functionality not present in the connector, you need to copy the original first

**Reply by xPaw** [reply]
Guess I am sticking prepending a minus before calling `refine` for now, it mostly works fine, except the remove button in `CurrentRefinements` doesn't work.
